src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -25,13 +47,234 @@ def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestPro
+
+    def add_article_to_localities(self, rec_localities, article):
+        localities = extract_locality(article)
+        for local in localities:
+            rec_localities[local] = rec_localities.get(local, 0) + 1
+
+    def normalized_categories_with_candidate(self, rec_categories, article):
+        rec_categories_with_candidate = rec_categories.copy()
+        self.add_article_to_categories(rec_categories_with_candidate, article)
+        return normalized_category_count(rec_categories_with_candidate)
+
+    def normalized_localities_with_candidate(self, rec_localities, article):
+        rec_localities_with_candidate = rec_localities.copy()
+        self.add_article_to_localities(rec_localities_with_candidate, article)
+        return normalized_category_count(rec_localities_with_candidate)
+
+    def calibration(
+        self, relevance_scores, articles, topic_preferences, locality_preferences, theta_topic, theta_local, topk
+    ) -> list[Article]:
+        # MR_i = (1 - theta_topic - theta_local) * reward_i - theta_topic * C_topic - theta_local * C_local
+        # R is all candidates (not selected yet)
+
+        recommendations = []  # final recommendation (topk index)
+
+        topic_categories = defaultdict(int)
+        local_categories = defaultdict(int)
+
+        for _ in range(topk):
+            candidate = None  # next item
+            best_candidate_score = float("-inf")
+
+            for article_idx, article_score in enumerate(relevance_scores):  # iterate R for next item
+                if article_idx in recommendations:
+                    continue
+
+                normalized_candidate_topics = self.normalized_categories_with_candidate(
+                    topic_categories, articles[article_idx]
+                )
+                normalized_candidate_locality = self.normalized_localities_with_candidate(
+                    local_categories, articles[article_idx]
+                )
+
+                calibration_topic = compute_kl_divergence(topic_preferences, normalized_candidate_topics)
+                calibration_local = compute_kl_divergence(locality_preferences, normalized_candidate_locality)
+
+                adjusted_candidate_score = (
+                    (1 - theta_local - theta_topic) * article_score
+                    - (theta_topic * calibration_topic)
+                    - (theta_local * calibration_local)
+                )
+                if adjusted_candidate_score > best_candidate_score:
+                    best_candidate_score = adjusted_candidate_score
+                    candidate = article_idx
+
+            if candidate is not None:
+                recommendations.append(candidate)
+                self.add_article_to_categories(topic_categories, articles[candidate])
+                self.add_article_to_localities(local_categories, articles[candidate])
+
+        return recommendations
+
+    def compute_local_dist(self, candidate_articles: ArticleSet):
```

Comment:
At first I thought this was computing a distance before figuring out from the return value that it's forming a distribution. Maybe it would be clearer if it was named with preferences in mind?
```suggestion
    def compute_local_prefs(self, candidate_articles: ArticleSet):
```